### ISSUE TITLE
Possibly wrong implementation of focal loss in regard of label smoothing

### ISSUE THREAD

**Issue by atuleu** [keras-team/keras-cv#672]
I think that there is a wrong implementation of focal loss in regard of label smoothing.

Here is the current implementation
https://github.com/keras-team/keras-cv/blob/35b9f9534fc193e00b4fa850963a05b331467065/keras_cv/losses/focal.py#L74-L90

as seen above, label smoothing occurs *before* the computation of `alpha` and `pt`. From the original paper, the value expected at this stage for `y_true` should be either `0` or `1`. However after label smoothing, `y_true` will be in `]0,1[`, therefore invalidating `alpha` and `pt` computation (at least not as written in the original paper).

Also to compute Multi-Label, I think the final reduction should be a sum to be correct. One alternative would be, as implemented in pytorch to add a `reduction` parameter, to either:

* 'none': perform no class reduction
* 'sum': perform a sum over classes
* 'mean': perform a mean over classes

What do you thin @LukeWood , @quantumalaviya ?

**Comment by atuleu** [comment]
Here are the pytorch and facebookresearch implementations:

https://pytorch.org/vision/stable/_modules/torchvision/ops/focal_loss.html
https://github.com/facebookresearch/fvcore/blob/main/fvcore/nn/focal_loss.py

**Comment by atuleu** [comment]
Possible fix available here https://github.com/atuleu/keras-cv/commit/ba35650bab435dcfe04000b0b7e68b7058829839 . Please tell me if I should open a PR

**Comment by quantumalaviya** [comment]
I agree, I think the label smoothing issue should be fixed. Thanks for pointing it out.

As for the reduction, `keras.losses.loss` already takes in a reduction argument which is handled by kwargs. No need to add a new argument there.

**Comment by atuleu** [comment]
I think it is not the same argument. Loss reduction in keras is per batch, to handle multiple device training properly. Here we are discussing of class reduction of the binary cross-entropy to implement a multi label / categorical cross entropy from binary cross entropy which should be handled at the FocalLoss level IMO.

**Comment by LukeWood** [comment]
Looks like a legitimate bug!  Thanks for the report @atuleu.

@quantumalaviya would you mind introducing a fix?

**Comment by LukeWood** [comment]
@atuleu lets do reduction in a separate issue!  one issue per thread is the easiest to discuss

**Comment by tanzhenyu** [comment]
@atuleu Marking this as contribution welcome. Please help us by sending a PR :-)

**Comment by DavidLandup0** [comment]
This was fixed with https://github.com/keras-team/keras-cv/pull/680
This issue should be closed :)

**Comment by bhack** [comment]
> @atuleu Marking this as contribution welcome. Please help us by sending a PR :-)

What do you mean? We have one since August:
 https://github.com/keras-team/keras-cv/pull/680

**Comment by DavidLandup0** [comment]
Changes were requested and it's still WIP
